scripts/generate_package_docs.py

```diff
@@ -0,0 +1,98 @@
+  file_name = "{}-{}-{}".format(params.image_dir, params.image_config['version'], params.config['doc_suffix'])
+  utils.write_json_to_file(docs, file_name)
+  utils.gsutil_cp(file_name, params.config["doc_bucket"])
+
+def get_docs(params):
+  tools = params.image_config["tools"]
+  doc_builder = get_doc_builder()
+
+  docs = {}
+  for tool in tools:
+    print "Writing doc for tool {}".format(tool)
+      #flattens the arrays of objects with packages:version key value pairs into single object with all key value pairs
+    docs[tool] = utils.flatten_list_of_dicts(doc_builder[tool](params))
+
+  return docs
+
+def get_doc_builder():
+  builder = {
+    "r": get_r_doc,
+    "python": get_python_doc,
+    "gatk": get_gatk_doc
+  }
+
+  return builder
+
+def get_r_doc(params):
+  raw_r_doc =  utils.docker_exec(params.image_dir, "sh -c \"echo 'ip <- as.data.frame(installed.packages()[,c(1,3:4)]) \n rownames(ip) <- NULL \n ip <- ip[is.na(ip\$Priority),1:2,drop=FALSE] \n print(ip, row.names=FALSE)' > temp.r; Rscript temp.r; rm temp.r\"")
+  packages = raw_r_doc.strip().split("\n")
+
+  r_version_output = utils.docker_exec(params.image_dir, "R --version | head -1")
+  try:
+    r_version = re.search("(\d+\.)(\d+\.)(\*|\d+)", r_version_output).group(0)
+  except AttributeError as e:
+    raise ValueError("the R has produced output with an improperly formatted version. Something has likely gone wrong with this script, ensure that the output has a version properly parsed by the regex in the function write_r_doc.\n\tOutput: " + r_version_output)
+
+  package_json = [{"r": r_version}]
+  for package in packages[1:]:
+    processed_tuple = re.sub(' +', ' ',package.strip()).split(' ')
+
+    assert len(processed_tuple) == 2,"Asserting format for {} is correct".format(package)
+     
+    json = { processed_tuple[0]: processed_tuple[1]}
+    package_json.append(json)
+
+  return package_json
+
+def get_python_doc(params):
+  python_version_output = utils.docker_exec(params.image_dir, "python3 --version")
+  raw_python_doc = utils.docker_exec(params.image_dir, "pip list --format=json")
+
+  try:
+    python_version = re.search("(\d+\.)(\d+\.)(\*|\d+)", python_version_output).group(0)
+  except AttributeError as e:
+    raise ValueError("the python command has produced output with an improperly formatted version. Something has likely gone wrong with this script, ensure that the output has a version properly parsed by the regex in the function write_python_doc. \n\tOutput: " + python_version_output)
+
+  json_python_doc = json.loads(raw_python_doc)
+  json_python_doc = list(map(lambda package: { package["name"]: package["version"] }, json_python_doc))
+  json_python_doc.append({"python": python_version})
+
+  return json_python_doc
+
+def get_gatk_doc(params):
+  cmd = "ls /etc/ | grep gatk | tr '-' ' ' | awk '{print $2}'"
+  gatk_version = utils.docker_exec(params.image_dir, cmd)
+
+  return [{"gatk": gatk_version}]
+
+class ParamProcessor():
```

Comment:
Instead of rolling own, it may be better to use a library. I used [ArgumentParser](http://zetcode.com/python/argparse/) on another library and it worked fairly well. It's the most helpful when centrally listing the args in a separate file. I understand that these are scripts so I'll let you judge the benefit of it  😊

Reply:
ooooh I didn't know about this built-in module this is great. I've always stayed away from complex flags, but now I'm itching for a use case. I think I'm going to leave it in this particular case, since it's fairly basic and there's no optional params, plus a bit of custom processing is needed with the params. Good callout though, I'll definitely keep this in mind over the custom ones I've been rolling.